carrot+fcmp: send to integrated address and fix transfer tests

jeffro256 opened the pull request:
Depends on #54

jeffro256 committed:
carrot+fcmp: send to integrated address and fix transfer tests

j-berman commented on tests/functional_tests/transfer.py:
```diff
@@ -1648,43 +1637,6 @@ def assert_correct_transfers(wallet, expected_transfers, expected_inc_transfers,
         receiver_wallet.refresh()
         assert_correct_transfers(receiver_wallet, transfers, incoming_transfers, expected_receiver_balance)
 
-        # Send after background sync
```
I think accidental delete? This is a useful test

jeffro256 replied:
Yes, sorry, this was accidental. To start out on this file, I reverted the changes in #50 to this file.

j-berman commented on tests/functional_tests/transfer.py:
```diff
@@ -73,7 +72,6 @@ def run_test(self):
         self.create()
         self.mine()
         self.transfer()
-        # TODO: self.bad_transfer() (test error handling, e.g. no unlocked outputs triggering "no input candidates provided")
```
I think we want this new test added to make sure the RPC wallet handles expected transfer errors correctly. Can be done in a separate PR, but I think it would be good to keep the TODO

jeffro256 replied:
Sorry, this was also accidentally kept in the reversion. It's good to keep the TODO. Latest commit puts it back in.

j-berman commented on tests/functional_tests/transfer.py:
```diff
@@ -1076,9 +1067,7 @@ def check_multiple_submissions(self):
-        # Needs https://github.com/seraphis-migration/monero/pull/49
-        print('TODO: FCMP++ Testing scan_tx')
-        return
+        print('Testing scan_tx')
```
`scan_tx` needs https://github.com/seraphis-migration/monero/pull/49 to work correctly fwiw, even if the test passes (I should add a test to send after scan also)

j-berman commented on src/wallet/wallet2.cpp:
```diff
@@ -7548,7 +7520,8 @@ void wallet2::commit_tx(pending_tx& ptx)
   if (store_tx_info())
   {
-    payment_id = get_payment_id(ptx);
+    const crypto::hash8 payment_id_8 = wallet::get_pending_tx_payment_id(ptx);
```
`get_pending_tx_payment_id` doesn't decrypt the payment ID for `PreCarrotTransactionProposal`'s, but `get_payment_id` did

jeffro256 replied:
Ah shoot, I didn't upstream the changes to constructing `PreCarrotTransactionProposal`'s in `wallet2` that makes this correct. You're very right, as-is, this isn't correct since this will grab the encrypted PID.

jeffro256 replied:
Okay this should be done now. Obviously cold wallets and multisig is still broken though...

jeffro256 committed:
carrot_core: make base exception inheritance public

Derived exceptions do not get caught as a base exception if the inheritance isn't public. TIL.

jeffro256 committed:
carrot_core: continue fixiing exception inheritance

j-berman commented on src/simplewallet/simplewallet.cpp:
```diff
@@ -7075,14 +7067,13 @@ bool simple_wallet::sweep_main(uint32_t account, uint64_t below, const std::vect
-    crypto::hash payment_id;
-    bool r = tools::wallet2::parse_long_payment_id(payment_id_str, payment_id);
+    crypto::hash ignored_long_pid;
+    bool r = tools::wallet2::parse_long_payment_id(payment_id_str, ignored_long_pid);
```
Hmm, so this means we're no longer going to use the long payment ID at all if it's provided. That's a significant change. How much of a pain would it be to support the long payment ID type?

j-berman replied:
Nevermind, even [the scanner ignores the long payment ID's](https://github.com/monero-project/monero/blob/c572e1ad00802fc83b756460838d01436a512750/src/wallet/wallet2.cpp#L2828-L2832). I think this is fine.

j-berman commented on src/simplewallet/simplewallet.cpp:
```diff
@@ -7561,14 +7502,18 @@ bool simple_wallet::donate(const std::vector<std::string> &args_)
+  if (payment_id != crypto::null_hash)
+  {
+    LONG_PAYMENT_ID_SUPPORT_CHECK();
+  }
```
Nit: doesn't look necessary since `transfer` would also print this presumably

jeffro256 replied:
Done

j-berman commented on src/simplewallet/simplewallet.cpp:
```diff
@@ -6492,8 +6492,7 @@ bool simple_wallet::transfer_main(const std::vector<std::string> &args_, bool ca
 
-  std::vector<uint8_t> extra;
-  bool payment_id_seen = false;
+  // Fail on parsed long payment ID
```
It's not failing though, just warning and ignoring, no?

jeffro256 replied:
`LONG_PAYMENT_ID_SUPPORT_CHECK()` logs a message in the failure format, then terminates with `return true;`. The transfer shouldn't continue after that AFAIK.

j-berman commented on src/wallet/api/wallet.cpp:
```diff
@@ -1643,15 +1641,11 @@ PendingTransaction *WalletImpl::createTransactionMultDest(const std::vector<stri
-                break;
-            }
+            setStatusError(tr("Long payment IDs are obsolete and no longer supported"));
+            break;
```
Thoughts on just getting rid of the `payment_id` param?

jeffro256 replied:
I'm a bit more iffy on breaking changes to `libwallet_api`...

j-berman commented on src/wallet/wallet2.cpp:
```diff
@@ -10420,24 +10393,29 @@ static uint32_t get_count_above(const std::vector<wallet2::transfer_details> &tr
   boost::unique_lock<hw::device> hwdev_lock (hwdev);
   hw::reset_mode rst(hwdev);  
 
+  check_tx_extra_construction_parameter(extra);
```
Just calling out, I think this is something we should document explicitly somewhere: we don't expect API callers to pass the payment ID in tx extra anymore, and to pass it via the `payment_id` param instead

jeffro256 replied:
Done. Added documentation to method declaration.

j-berman commented on src/wallet/tx_builder.cpp:
```diff
@@ -143,13 +144,15 @@ static void build_payment_proposals_sweep(std::vector<carrot::CarrotPaymentPropo
         build_payment_proposals(normal_payment_proposals_inout,
             selfsend_payment_proposals_inout,
             cryptonote::tx_destination_entry(0, address, is_subaddress),
+            i ? crypto::null_hash8 : payment_id,
```
Just a passing question: is the order of dest enotes in the final tx shuffled?

jeffro256 replied:
It's sorted deterministically by one-time address. This lines assigns the payment ID to *one of* the output enotes (not externally visible). The addresses for all the enotes are all the same in a sweep tx. It's a bit weird in that a "random" enote will have the PID attached to it, and the others won't, but I don't think that this really leaks anything meaningful...

jeffro256 committed:
carrot+fcmp: revert previous overzealous reversion in transfer.py

jeffro256 committed:
transfer.py re-import math

jeffro256 committed:
carrot+fcmp: revert print statement changes in transfer.py in #50

jeffro256 committed:
carrot+fcmp: @j-berman SW review: long PID in `donate` handled in `transfer`

jeffro256 committed:
carrot+fcmp: PID @j-berman review: set construction extra w/ decrypted PID

jeffro256 commented:
Only test which failed was the `select_outputs.density` unit test, which fails sporadically with bad entropy.
